src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +43,209 @@ public class Settings : ISettings
             new[] { "*.config" } :
             new[] { "*.Config", "*.config" };
 
-        private XDocument ConfigXDocument { get; }
-        public string FileName { get; }
-        private bool IsMachineWideSettings { get; }
-        // next config file to read if any
-        private Settings _next;
-        // The priority of this setting file
-        private int _priority;
+        private SettingsFile _settingsHead { get; }
 
-        private bool Cleared { get; set; }
+        private Dictionary<string, AbstractSettingSection> _computedSections { get; set; }
 
-        public Settings(string root /*, ILogger logger */)
-            : this(root, DefaultSettingsFileName, false)
+        public SettingSection GetSection(string sectionName)
         {
+            if (_computedSections.TryGetValue(sectionName, out var section))
+            {
+                return section.Clone() as SettingSection;
+            }
+
+            return null;
         }
 
-        public Settings(string root, string fileName)
-            : this(root, fileName, false)
+        public void AddOrUpdate(string sectionName, SettingItem item)
         {
+            if (string.IsNullOrEmpty(sectionName))
+            {
+                throw new ArgumentNullException(nameof(sectionName));
+            }
+
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            Debug.Assert(item.IsAbstract());
+
+            // Operation is an update
+            if (_computedSections.TryGetValue(sectionName, out var section) && section.Items.Contains(item))
+            {
+                // An update could not be possible here because the operation might be
+                // in a machine wide config. If so then we want to add the item to
+                // the output config.
+                if (section.Update(item))
+                {
+                    return;
+                }
+            }
+
+            // Operation is an add
+            var outputSettingsFile = GetOutputSettingFileForSection(sectionName);
+            if (outputSettingsFile != null)
```

Comment:
We could throw. This basically means that if there is no writeable file we ignore the update. I opted to ignore instead of throw to keep the same behavior as before, but I'm in favor of throwing to have a better error experience.